File: transform/mapper.py
"""
Data transformation from Airtable records to HSDS models.

Maps Airtable field names to HSDS schema fields.
"""
from typing import Dict, Any, List, Optional
from models.hsds import (
    Organization, OrganizationSummary, Service, ServiceSummary,
    Location, ServiceAtLocation, Taxonomy, TaxonomyTerm,
    Phone, Address, Contact, Schedule, Language, ServiceArea,
    Program, Funding, CostOption, RequiredDocument, Accessibility,
    Page
)


def safe_float(value: Any) -> Optional[float]:
    """Safely convert a value to float."""
    if value is None:
        return None
    try:
        # Handle string values (from multilineText fields)
        if isinstance(value, str):
            value = value.strip()
            if not value:
                return None
        return float(value)
    except (ValueError, TypeError):
        return None


def safe_int(value: Any) -> Optional[int]:
    """Safely convert a value to int."""
    if value is None:
        return None
    try:
        return int(value)
    except (ValueError, TypeError):
        return None


def first_or_none(lst: Optional[List]) -> Optional[Any]:
    """Get first element of list or None."""
    if lst and len(lst) > 0:
        return lst[0]
    return None


def join_list(lst: Optional[List]) -> Optional[str]:
    """Join list elements with comma."""
    if lst:
        return ", ".join(str(x) for x in lst)
    return None


class HSDSMapper:
    """Maps Airtable records to HSDS Pydantic models."""
    
    def __init__(self, record_cache: Dict[str, Dict[str, Any]] = None):
        """
        Initialize mapper with optional record cache.
        
        The cache maps table names to dicts of record_id -> fields,
        used for resolving linked records.
        """
        self.cache = record_cache or {}
    
    def map_phone(self, data: Dict[str, Any]) -> Phone:
        """Map Airtable phone record to HSDS Phone."""
        return Phone(
            id=data.get("id", ""),
            number=data.get("number", ""),
            extension=data.get("extension"),
            type=data.get("type"),
            description=data.get("description"),
        )
    
    def map_address(self, data: Dict[str, Any]) -> Address:
        """Map Airtable address record to HSDS Address."""
        # Handle address_type as list or string
        address_type = data.get("address_type")
        if isinstance(address_type, list):
            address_type = first_or_none(address_type)
        
        return Address(
            id=data.get("id", ""),
            address_1=data.get("address_1"),
            address_2=data.get("address_2"),
            city=data.get("city"),
            state_province=data.get("state_province"),
            postal_code=data.get("postal_code"),
            region=data.get("region"),
            country=data.get("country"),
            address_type=address_type,
            attention=data.get("attention"),
        )
    
    def map_language(self, data: Dict[str, Any]) -> Language:
        """Map Airtable language record to HSDS Language."""
        return Language(
            id=data.get("id", ""),
            name=data.get("name"),
            code=data.get("code"),
            note=data.get("note"),
        )
    
    def map_accessibility(self, data: Dict[str, Any]) -> Accessibility:
        """Map Airtable accessibility record to HSDS Accessibility."""
        return Accessibility(
            id=data.get("id", ""),
            description=data.get("description"),
            details=data.get("details"),
            url=data.get("url"),
        )
    
    def map_schedule(self, data: Dict[str, Any]) -> Schedule:
        """Map Airtable schedule record to HSDS Schedule."""
        # Handle byday as list
        byday = data.get("byday")
        if isinstance(byday, list):
            byday = ",".join(byday)
        
        return Schedule(
            id=data.get("id", ""),
            valid_from=data.get("valid_from"),
            valid_to=data.get("valid_to"),
            dtstart=data.get("dtstart"),
            timezone=data.get("timezone"),
            until=data.get("until"),
            count=data.get("count"),
            wkst=data.get("wkst"),
            freq=data.get("freq"),
            interval=data.get("interval"),
            byday=byday,
            byweekno=data.get("byweekno"),
            bymonthday=data.get("bymonthday"),
            byyearday=data.get("byyearday"),
            description=data.get("description"),
            opens_at=data.get("opens_at"),
            closes_at=data.get("closes_at"),
            schedule_link=data.get("schedule_link"),
            attending_type=data.get("attending_type"),
            notes=data.get("notes"),
        )
    
    def map_contact(self, data: Dict[str, Any], phones: List[Phone] = None) -> Contact:
        """Map Airtable contact record to HSDS Contact."""
        return Contact(
            id=data.get("id", ""),
            name=data.get("name"),
            title=data.get("title"),
            department=data.get("department"),
            email=data.get("email"),
            phones=phones,
        )
    
    def map_service_area(self, data: Dict[str, Any]) -> ServiceArea:
        """Map Airtable service_area record to HSDS ServiceArea."""
        return ServiceArea(
            id=data.get("id", ""),
            name=data.get("name"),
            description=data.get("description"),
            extent=data.get("extent"),
            extent_type=data.get("extent_type"),
            uri=data.get("uri"),
        )
    
    def map_program(self, data: Dict[str, Any]) -> Program:
        """Map Airtable program record to HSDS Program."""
        return Program(
            id=data.get("id", ""),
            name=data.get("name", ""),
            alternate_name=data.get("alternate_name"),
            description=data.get("description"),
        )
    
    def map_funding(self, data: Dict[str, Any]) -> Funding:
        """Map Airtable funding record to HSDS Funding."""
        return Funding(
            id=data.get("id", ""),
            source=data.get("source"),
        )
    
    def map_cost_option(self, data: Dict[str, Any]) -> CostOption:
        """Map Airtable cost_option record to HSDS CostOption."""
        return CostOption(
            id=data.get("id", ""),
            option=data.get("option"),
            currency=data.get("currency"),
            amount=safe_float(data.get("amount")),
            amount_description=data.get("amount_description"),
            valid_from=data.get("valid_from"),
            valid_to=data.get("valid_to"),
        )
    
    def map_required_document(self, data: Dict[str, Any]) -> RequiredDocument:
        """Map Airtable required_document record to HSDS RequiredDocument."""
        return RequiredDocument(
            id=data.get("id", ""),
            document=data.get("document"),
            uri=data.get("uri"),
        )
    
    def map_taxonomy(self, data: Dict[str, Any]) -> Taxonomy:
        """Map Airtable taxonomy record to HSDS Taxonomy."""
        return Taxonomy(
            id=data.get("id", ""),
            name=data.get("name", ""),
            description=data.get("description"),
            uri=data.get("uri"),
            version=data.get("version"),
        )
    
    def map_taxonomy_term(self, data: Dict[str, Any], taxonomy: Taxonomy = None) -> TaxonomyTerm:
        """Map Airtable taxonomy_term record to HSDS TaxonomyTerm."""
        return TaxonomyTerm(
            id=data.get("id", ""),
            name=data.get("name", ""),
            code=data.get("code"),
            description=data.get("description"),
            parent_id=first_or_none(data.get("parent")),
            taxonomy=first_or_none(data.get("taxonomy")),
            taxonomy_detail=taxonomy,
            language=data.get("language"),
            term_uri=data.get("term_uri"),
        )
    
    def map_location(
        self,
        data: Dict[str, Any],
        addresses: List[Address] = None,
        phones: List[Phone] = None,
        contacts: List[Contact] = None,
        accessibility: List[Accessibility] = None,
        languages: List[Language] = None,
        schedules: List[Schedule] = None,
    ) -> Location:
        """Map Airtable location record to HSDS Location."""
        # Handle location_type as list
        location_type = data.get("location_type")
        if isinstance(location_type, list):
            location_type = first_or_none(location_type)
        
        return Location(
            id=data.get("id", ""),
            location_type=location_type,
            url=data.get("url"),
            name=data.get("name"),
            alternate_name=data.get("alternate_name"),
            description=data.get("description"),
            transportation=data.get("transportation"),
            latitude=safe_float(data.get("latitude")),
            longitude=safe_float(data.get("longitude")),
            external_identifier=data.get("external_identifier"),
            external_identifier_type=data.get("external_identifier_type"),
            addresses=addresses,
            phones=phones,
            contacts=contacts,
            accessibility=accessibility,
            languages=languages,
            schedules=schedules,
        )
    
    def map_organization_summary(self, data: Dict[str, Any]) -> OrganizationSummary:
        """Map Airtable organization to minimal HSDS summary."""
        return OrganizationSummary(
            id=data.get("id", ""),
            name=data.get("name", ""),
            alternate_name=data.get("alternate_name"),
            description=data.get("description"),
            email=data.get("email"),
            website=data.get("website"),
            logo=data.get("logo"),
            uri=data.get("uri"),
        )
    
    def map_organization(
        self,
        data: Dict[str, Any],
        phones: List[Phone] = None,
        contacts: List[Contact] = None,
        locations: List[Location] = None,
        programs: List[Program] = None,
        funding: List[Funding] = None,
    ) -> Organization:
        """Map Airtable organization record to full HSDS Organization."""
        return Organization(
            id=data.get("id", ""),
            name=data.get("name", ""),
            alternate_name=data.get("alternate_name"),
            description=data.get("description"),
            email=data.get("email"),
            website=data.get("website"),
            year_incorporated=safe_int(data.get("year_incorporated")),
            legal_status=data.get("legal_status"),
            logo=data.get("logo"),
            uri=data.get("uri"),
            parent_organization_id=first_or_none(data.get("organization")),
            phones=phones,
            contacts=contacts,
            locations=locations,
            programs=programs,
            funding=funding,
        )
    
    def map_service_at_location(
        self,
        data: Dict[str, Any],
        location: Location = None,
        phones: List[Phone] = None,
        contacts: List[Contact] = None,
        schedules: List[Schedule] = None,
        service_areas: List[ServiceArea] = None,
    ) -> ServiceAtLocation:
        """Map Airtable service_at_location record to HSDS ServiceAtLocation."""
        return ServiceAtLocation(
            id=data.get("id", ""),
            service_id=first_or_none(data.get("services")),
            description=data.get("description"),
            location=location,
            phones=phones,
            contacts=contacts,
            schedules=schedules,
            service_areas=service_areas,
        )
    
    def map_service_summary(
        self,
        data: Dict[str, Any],
        organization_id: str,
        organization: OrganizationSummary = None,
        program: Program = None,
    ) -> ServiceSummary:
        """Map Airtable service to minimal HSDS summary (ORUK compliant)."""
        return ServiceSummary(
            id=data.get("id", ""),
            organization_id=organization_id,  # Required by ORUK
            name=data.get("name", ""),
            status=data.get("status") or "active",  # Required by ORUK
            alternate_name=data.get("alternate_name"),
            description=data.get("description"),
            url=data.get("url"),
            email=data.get("email"),
            last_modified=data.get("lastUpdated"),
            organization=organization,
            program=program,
            # Taxonomy fields from Airtable (camelCase in Airtable)
            need_focus=data.get("needFocus") or [],
            community_focus=data.get("communityFocus") or [],
        )
    
    def map_service(
        self,
        data: Dict[str, Any],
        organization_id: str,
        organization: OrganizationSummary = None,
        program: Program = None,
        phones: List[Phone] = None,
        contacts: List[Contact] = None,
        schedules: List[Schedule] = None,
        service_areas: List[ServiceArea] = None,
        service_at_locations: List[ServiceAtLocation] = None,
        languages: List[Language] = None,
        funding: List[Funding] = None,
        cost_options: List[CostOption] = None,
        required_documents: List[RequiredDocument] = None,
    ) -> Service:
        """Map Airtable service record to full HSDS Service (ORUK compliant)."""
        # Extract group name from Airtable's groupName field (lookup field returns list)
        group_name_list = data.get("groupName", [])
        group_name = group_name_list[0] if group_name_list else None
        
        return Service(
            id=data.get("id", ""),
            organization_id=organization_id,  # Required by ORUK
            name=data.get("name", ""),
            status=data.get("status") or "active",  # Required by ORUK
            alternate_name=data.get("alternate_name"),
            description=data.get("description"),
            url=data.get("url"),
            email=data.get("email"),
            interpretation_services=data.get("interpretation_services"),
            application_process=data.get("application_process"),
            fees_description=data.get("fees_description"),
            accreditations=data.get("accreditations"),
            eligibility_description=data.get("eligibility_description"),
            minimum_age=safe_int(data.get("minimum_age")),
            maximum_age=safe_int(data.get("maximum_age")),
            assured_date=data.get("assured_date"),
            assurer_email=data.get("assurer_email"),
            alert=data.get("alert"),
            last_modified=data.get("lastUpdated"),
            organization=organization,
            program=program,
            phones=phones,
            contacts=contacts,
            schedules=schedules,
            service_areas=service_areas,
            service_at_locations=service_at_locations,
            languages=languages,
            funding=funding,
            cost_options=cost_options,
            required_documents=required_documents,
            # Custom extension fields
            group_name=group_name,
            need_focus=data.get("needFocus"),
            community_focus=data.get("communityFocus"),
        )
    
    @staticmethod
    def paginate(items: List, page: int, per_page: int) -> Page:
        """Create a paginated response."""
        total_items = len(items)
        total_pages = max(1, (total_items + per_page - 1) // per_page)
        
        start = (page - 1) * per_page
        end = start + per_page
        page_items = items[start:end]
        
        return Page(
            total_items=total_items,
            total_pages=total_pages,
            page_number=page,
            size=len(page_items),
            first_page=(page == 1),
            last_page=(page >= total_pages),
            empty=(len(page_items) == 0),
            contents=page_items,
        )

File: worker/src/mapper.ts
/**
 * Airtable → HSDS field mapper.
 *
 * Ports transform/mapper.py to TypeScript.
 * Pure functions that transform raw JSON data objects into typed HSDS shapes.
 */
import type {
  Phone, Address, Language, Accessibility, Schedule, Contact,
  ServiceArea, Program, Funding, CostOption, RequiredDocument,
  Taxonomy, TaxonomyTerm, Location, OrganizationSummary, Organization,
  ServiceAtLocation, ServiceSummary, Service, Page,
} from "./types";

// ============================================================================
// Utility helpers
// ============================================================================

export function safeFloat(value: unknown): number | undefined {
  if (value == null) return undefined;
  if (typeof value === "string") {
    const trimmed = value.trim();
    if (!trimmed) return undefined;
    const n = Number(trimmed);
    return Number.isNaN(n) ? undefined : n;
  }
  if (typeof value === "number") return value;
  return undefined;
}

export function safeInt(value: unknown): number | undefined {
  if (value == null) return undefined;
  const n = Number(value);
  return Number.isNaN(n) ? undefined : Math.floor(n);
}

export function firstOrNone<T>(list: T[] | undefined | null): T | undefined {
  return list && list.length > 0 ? list[0] : undefined;
}

export function joinList(list: unknown[] | undefined | null): string | undefined {
  if (!list || list.length === 0) return undefined;
  return list.map(String).join(", ");
}

/**
 * Remove undefined/null values from an object (ORUK compliance —
 * optional fields without values should be omitted).
 */
export function stripNulls<T extends Record<string, unknown>>(obj: T): Partial<T> {
  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(obj)) {
    if (value !== undefined && value !== null) {
      result[key] = value;
    }
  }
  return result as Partial<T>;
}

// ============================================================================
// Map functions — each mirrors a Python map_* method
// ============================================================================

export function mapPhone(data: Record<string, unknown>): Phone {
  return stripNulls({
    id: (data.id as string) || "",
    number: (data.number as string) || "",
    extension: data.extension as string | undefined,
    type: data.type as string | undefined,
    description: data.description as string | undefined,
  }) as Phone;
}

export function mapAddress(data: Record<string, unknown>): Address {
  let addressType = data.address_type;
  if (Array.isArray(addressType)) addressType = firstOrNone(addressType);

  return stripNulls({
    id: (data.id as string) || "",
    address_1: data.address_1 as string | undefined,
    address_2: data.address_2 as string | undefined,
    city: data.city as string | undefined,
    state_province: data.state_province as string | undefined,
    postal_code: data.postal_code as string | undefined,
    region: data.region as string | undefined,
    country: data.country as string | undefined,
    address_type: addressType as string | undefined,
    attention: data.attention as string | undefined,
  }) as Address;
}

export function mapLanguage(data: Record<string, unknown>): Language {
  return stripNulls({
    id: (data.id as string) || "",
    name: data.name as string | undefined,
    code: data.code as string | undefined,
    note: data.note as string | undefined,
  }) as Language;
}

export function mapAccessibility(data: Record<string, unknown>): Accessibility {
  return stripNulls({
    id: (data.id as string) || "",
    description: data.description as string | undefined,
    details: data.details as string | undefined,
    url: data.url as string | undefined,
  }) as Accessibility;
}

export function mapSchedule(data: Record<string, unknown>): Schedule {
  let byday = data.byday;
  if (Array.isArray(byday)) byday = byday.join(",");

  return stripNulls({
    id: (data.id as string) || "",
    valid_from: data.valid_from as string | undefined,
    valid_to: data.valid_to as string | undefined,
    dtstart: data.dtstart as string | undefined,
    timezone: data.timezone as string | undefined,
    until: data.until as string | undefined,
    count: data.count as string | undefined,
    wkst: data.wkst as string | undefined,
    freq: data.freq as string | undefined,
    interval: data.interval as string | undefined,
    byday: byday as string | undefined,
    byweekno: data.byweekno as string | undefined,
    bymonthday: data.bymonthday as string | undefined,
    byyearday: data.byyearday as string | undefined,
    description: data.description as string | undefined,
    opens_at: data.opens_at as string | undefined,
    closes_at: data.closes_at as string | undefined,
    schedule_link: data.schedule_link as string | undefined,
    attending_type: data.attending_type as string | undefined,
    notes: data.notes as string | undefined,
  }) as Schedule;
}

export function mapContact(
  data: Record<string, unknown>,
  phones?: Phone[],
): Contact {
  return stripNulls({
    id: (data.id as string) || "",
    name: data.name as string | undefined,
    title: data.title as string | undefined,
    department: data.department as string | undefined,
    email: data.email as string | undefined,
    phones: phones && phones.length > 0 ? phones : undefined,
  }) as Contact;
}

export function mapServiceArea(data: Record<string, unknown>): ServiceArea {
  return stripNulls({
    id: (data.id as string) || "",
    name: data.name as string | undefined,
    description: data.description as string | undefined,
    extent: data.extent as string | undefined,
    extent_type: data.extent_type as string | undefined,
    uri: data.uri as string | undefined,
  }) as ServiceArea;
}

export function mapProgram(data: Record<string, unknown>): Program {
  return stripNulls({
    id: (data.id as string) || "",
    name: (data.name as string) || "",
    alternate_name: data.alternate_name as string | undefined,
    description: data.description as string | undefined,
  }) as Program;
}

export function mapFunding(data: Record<string, unknown>): Funding {
  return stripNulls({
    id: (data.id as string) || "",
    source: data.source as string | undefined,
  }) as Funding;
}

export function mapCostOption(data: Record<string, unknown>): CostOption {
  return stripNulls({
    id: (data.id as string) || "",
    option: data.option as string | undefined,
    currency: data.currency as string | undefined,
    amount: safeFloat(data.amount),
    amount_description: data.amount_description as string | undefined,
    valid_from: data.valid_from as string | undefined,
    valid_to: data.valid_to as string | undefined,
  }) as CostOption;
}

export function mapRequiredDocument(data: Record<string, unknown>): RequiredDocument {
  return stripNulls({
    id: (data.id as string) || "",
    document: data.document as string | undefined,
    uri: data.uri as string | undefined,
  }) as RequiredDocument;
}

export function mapTaxonomy(data: Record<string, unknown>): Taxonomy {
  return stripNulls({
    id: (data.id as string) || "",
    name: (data.name as string) || "",
    description: data.description as string | undefined,
    uri: data.uri as string | undefined,
    version: data.version as string | undefined,
  }) as Taxonomy;
}

export function mapTaxonomyTerm(
  data: Record<string, unknown>,
  taxonomy?: Taxonomy,
): TaxonomyTerm {
  return stripNulls({
    id: (data.id as string) || "",
    name: (data.name as string) || "",
    code: data.code as string | undefined,
    description: data.description as string | undefined,
    parent_id: firstOrNone(data.parent as string[] | undefined),
    taxonomy: firstOrNone(data.taxonomy as string[] | undefined),
    taxonomy_detail: taxonomy,
    language: data.language as string | undefined,
    term_uri: data.term_uri as string | undefined,
  }) as TaxonomyTerm;
}

export function mapLocation(
  data: Record<string, unknown>,
  opts?: {
    addresses?: Address[];
    phones?: Phone[];
    contacts?: Contact[];
    accessibility?: Accessibility[];
    languages?: Language[];
    schedules?: Schedule[];
  },
): Location {
  let locationType = data.location_type;
  if (Array.isArray(locationType)) locationType = firstOrNone(locationType);

  return stripNulls({
    id: (data.id as string) || "",
    location_type: locationType as string | undefined,
    url: data.url as string | undefined,
    name: data.name as string | undefined,
    alternate_name: data.alternate_name as string | undefined,
    description: data.description as string | undefined,
    transportation: data.transportation as string | undefined,
    latitude: safeFloat(data.latitude),
    longitude: safeFloat(data.longitude),
    external_identifier: data.external_identifier as string | undefined,
    external_identifier_type: data.external_identifier_type as string | undefined,
    addresses: opts?.addresses && opts.addresses.length > 0 ? opts.addresses : undefined,
    phones: opts?.phones && opts.phones.length > 0 ? opts.phones : undefined,
    contacts: opts?.contacts && opts.contacts.length > 0 ? opts.contacts : undefined,
    accessibility: opts?.accessibility && opts.accessibility.length > 0 ? opts.accessibility : undefined,
    languages: opts?.languages && opts.languages.length > 0 ? opts.languages : undefined,
    schedules: opts?.schedules && opts.schedules.length > 0 ? opts.schedules : undefined,
  }) as Location;
}

export function mapOrganizationSummary(data: Record<string, unknown>): OrganizationSummary {
  return stripNulls({
    id: (data.id as string) || "",
    name: (data.name as string) || "",
    alternate_name: data.alternate_name as string | undefined,
    description: data.description as string | undefined,
    email: data.email as string | undefined,
    website: data.website as string | undefined,
    logo: data.logo as string | undefined,
    uri: data.uri as string | undefined,
  }) as OrganizationSummary;
}

export function mapOrganization(
  data: Record<string, unknown>,
  opts?: {
    phones?: Phone[];
    contacts?: Contact[];
    locations?: Location[];
    programs?: Program[];
    funding?: Funding[];
  },
): Organization {
  return stripNulls({
    id: (data.id as string) || "",
    name: (data.name as string) || "",
    alternate_name: data.alternate_name as string | undefined,
    description: data.description as string | undefined,
    email: data.email as string | undefined,
    website: data.website as string | undefined,
    year_incorporated: safeInt(data.year_incorporated),
    legal_status: data.legal_status as string | undefined,
    logo: data.logo as string | undefined,
    uri: data.uri as string | undefined,
    parent_organization_id: firstOrNone(data.organization as string[] | undefined),
    phones: opts?.phones && opts.phones.length > 0 ? opts.phones : undefined,
    contacts: opts?.contacts && opts.contacts.length > 0 ? opts.contacts : undefined,
    locations: opts?.locations && opts.locations.length > 0 ? opts.locations : undefined,
    programs: opts?.programs && opts.programs.length > 0 ? opts.programs : undefined,
    funding: opts?.funding && opts.funding.length > 0 ? opts.funding : undefined,
  }) as Organization;
}

export function mapServiceAtLocation(
  data: Record<string, unknown>,
  opts?: {
    location?: Location;
    phones?: Phone[];
    contacts?: Contact[];
    schedules?: Schedule[];
    service_areas?: ServiceArea[];
  },
): ServiceAtLocation {
  return stripNulls({
    id: (data.id as string) || "",
    service_id: firstOrNone(data.services as string[] | undefined),
    description: data.description as string | undefined,
    location: opts?.location,
    phones: opts?.phones && opts.phones.length > 0 ? opts.phones : undefined,
    contacts: opts?.contacts && opts.contacts.length > 0 ? opts.contacts : undefined,
    schedules: opts?.schedules && opts.schedules.length > 0 ? opts.schedules : undefined,
    service_areas: opts?.service_areas && opts.service_areas.length > 0 ? opts.service_areas : undefined,
  }) as ServiceAtLocation;
}

export function mapServiceSummary(
  data: Record<string, unknown>,
  organizationId: string,
  organization?: OrganizationSummary,
): ServiceSummary {
  return stripNulls({
    id: (data.id as string) || "",
    organization_id: organizationId,
    name: (data.name as string) || "",
    status: (data.status as string) || "active",
    alternate_name: data.alternate_name as string | undefined,
    description: data.description as string | undefined,
    url: data.url as string | undefined,
    email: data.email as string | undefined,
    last_modified: data.lastUpdated as string | undefined,
    organization,
    need_focus: (data.needFocus as string[]) || [],
    community_focus: (data.communityFocus as string[]) || [],
  }) as ServiceSummary;
}

export function mapService(
  data: Record<string, unknown>,
  organizationId: string,
  opts?: {
    organization?: OrganizationSummary;
    program?: Program;
    phones?: Phone[];
    contacts?: Contact[];
    schedules?: Schedule[];
    service_areas?: ServiceArea[];
    service_at_locations?: ServiceAtLocation[];
    languages?: Language[];
    funding?: Funding[];
    cost_options?: CostOption[];
    required_documents?: RequiredDocument[];
  },
): Service {
  const groupNameList = data.groupName as string[] | undefined;
  const groupName = groupNameList && groupNameList.length > 0 ? groupNameList[0] : undefined;

  return stripNulls({
    id: (data.id as string) || "",
    organization_id: organizationId,
    name: (data.name as string) || "",
    status: (data.status as string) || "active",
    alternate_name: data.alternate_name as string | undefined,
    description: data.description as string | undefined,
    url: data.url as string | undefined,
    email: data.email as string | undefined,
    interpretation_services: data.interpretation_services as string | undefined,
    application_process: data.application_process as string | undefined,
    fees_description: data.fees_description as string | undefined,
    accreditations: data.accreditations as string | undefined,
    eligibility_description: data.eligibility_description as string | undefined,
    minimum_age: safeInt(data.minimum_age),
    maximum_age: safeInt(data.maximum_age),
    assured_date: data.assured_date as string | undefined,
    assurer_email: data.assurer_email as string | undefined,
    alert: data.alert as string | undefined,
    last_modified: data.lastUpdated as string | undefined,
    organization: opts?.organization,
    program: opts?.program,
    phones: opts?.phones && opts.phones.length > 0 ? opts.phones : undefined,
    contacts: opts?.contacts && opts.contacts.length > 0 ? opts.contacts : undefined,
    schedules: opts?.schedules && opts.schedules.length > 0 ? opts.schedules : undefined,
    service_areas: opts?.service_areas && opts.service_areas.length > 0 ? opts.service_areas : undefined,
    service_at_locations: opts?.service_at_locations && opts.service_at_locations.length > 0 ? opts.service_at_locations : undefined,
    languages: opts?.languages && opts.languages.length > 0 ? opts.languages : undefined,
    funding: opts?.funding && opts.funding.length > 0 ? opts.funding : undefined,
    cost_options: opts?.cost_options && opts.cost_options.length > 0 ? opts.cost_options : undefined,
    required_documents: opts?.required_documents && opts.required_documents.length > 0 ? opts.required_documents : undefined,
    group_name: groupName,
    need_focus: data.needFocus as string[] | undefined,
    community_focus: data.communityFocus as string[] | undefined,
  }) as Service;
}

// ============================================================================
// Pagination helper
// ============================================================================

/**
 * Create a paginated Page response from total count and current page items.
 */
export function paginate<T>(
  items: T[],
  totalItems: number,
  page: number,
  perPage: number,
): Page<T> {
  const totalPages = Math.max(1, Math.ceil(totalItems / perPage));
  return {
    total_items: totalItems,
    total_pages: totalPages,
    page_number: page,
    size: items.length,
    first_page: page === 1,
    last_page: page >= totalPages,
    empty: items.length === 0,
    contents: items,
  };
}
